// @eil-protocol/sdk/dist/utils/AtomicSwapUtils.js
import { encodeFunctionData, keccak256, sliceHex, toBytes } from 'viem';
import AtomicSwapTypes from '@eil-protocol/contracts/artifacts/src/common/utils/AtomicSwapTypes.sol/AtomicSwapTypes.json' with { type: 'json' };
import { EmptySessionData } from '../contractTypes/SessionData.js';
export function getVoucherRequestId(atomicSwap) {
    const message = abiEncodeAtomicSwapForSigning(atomicSwap);
    return keccak256(toBytes(message));
}
export function abiEncodeVouchers(vouchers, sessionData = EmptySessionData) {
    const data = encodeFunctionData({
        abi: AtomicSwapTypes.abi,
        functionName: 'getVouchers',
        args: [vouchers, sessionData]
    });
    return sliceHex(data, 4);
}
/**
 * @notice Here only the 'data' bytes are passed to encoding, as the results will be used to create an ephemeral signature.
 * @param vouchers
 * @param sessionData
 */
export function abiEncodeVouchersForEphemeralSigning(vouchers, sessionData) {
    const data = encodeFunctionData({
        abi: AtomicSwapTypes.abi,
        functionName: 'getVouchersForEphemeralSigning',
        args: [vouchers, sessionData]
    });
    return sliceHex(data, 4);
}
/**
 * ABI-encode the {@link VoucherRequest} - the shared part of the paymaster data across layers.
 *
 * @notice On the destination chain the real paymaster data also includes the xlp's signed Voucher.
 * @param voucherRequest
 */
export function abiEncodeAtomicSwapForSigning(voucherRequest) {
    const enc = encodeFunctionData({
        abi: AtomicSwapTypes.abi,
        functionName: 'getVoucherRequest',
        args: [voucherRequest]
    });
    return sliceHex(enc, 4); // remove selector
}
/**
 * ABI-encode the entire Paymaster data for UserOperations on either layer.
 *
 * @param voucherRequestsDest
 */
export function abiEncodePaymasterData(voucherRequestsDest) {
    const data = encodeFunctionData({
        abi: AtomicSwapTypes.abi,
        functionName: 'getVoucherRequests',
        args: [voucherRequestsDest]
    });
    return sliceHex(data, 4); // remove selector
}
