// @eil-protocol/sdk/dist/utils/createClient.js
import { createPublicClient, createWalletClient, http } from 'viem';
import { privateKeyToAccount, toAccount } from 'viem/accounts';
import { getMasterAccount } from '../sdk/index.js';
function createChain(transport, chainId) {
    const client = createPublicClient({ transport });
    return {
        id: chainId,
        name: `Chain ${chainId}`,
        nativeCurrency: { name: 'Native', symbol: 'nat', decimals: 18 },
        rpcUrls: { default: { http: [client.transport.url] } }
    };
}
export function createWalletClientFromUrl(rpcUrl, chainId, account) {
    const transport = http(rpcUrl);
    return createWalletClient({
        account,
        chain: createChain(transport, chainId),
        transport,
    });
}
export function createPublicClientFromUrl(rpcUrl, chainId, pollingInterval) {
    const transport = http(rpcUrl);
    return createPublicClient({
        chain: createChain(transport, chainId),
        transport,
        pollingInterval
    });
}
export async function getChainIdFromUrl(rpcUrl) {
    return createPublicClientFromUrl(rpcUrl, 0).getChainId();
}
export async function createPublicClientWithChainIdFromUrl(rpcUrl) {
    const chainId = await getChainIdFromUrl(rpcUrl);
    const publicClient = createPublicClientFromUrl(rpcUrl, chainId);
    return publicClient;
}
export async function getClientsForDeployment(url) {
    const publicClient = await createPublicClientWithChainIdFromUrl(url);
    const chainId = await publicClient.getChainId();
    const deployerKey = process.env.DEPLOYER_KEY;
    const walletClient = deployerKey == null
        ? createWalletClientFromUrl(url, chainId, toAccount(await getMasterAccount(publicClient)))
        : createWalletClientFromUrl(url, chainId, privateKeyToAccount(deployerKey));
    return { publicClient, walletClient };
}
