// @eil-protocol/sdk/dist/utils/getSolventXlps.js
import { getContract } from 'viem';
import XlpSelectionHelperMeta from '@eil-protocol/contracts/artifacts/src/common/utils/XlpSelectionHelper.sol/XlpSelectionHelper.json' with { type: 'json' };
import { getCreate2DeployAddress } from '../deployments/create2deploy.js';
import { amountOrMinAmount, toAddress } from '../sdk/index.js';
export class SolventXlpSorter {
    arr;
    tokens;
    factor;
    constructor(arr, tokens, factor) {
        this.arr = arr;
        this.tokens = tokens;
        this.factor = factor;
    }
    // high priority: deposits ≥ factor × requested amount
    // medium: deposits ≥ requested amount
    // low: requires wallet balance top-up
    _sortPriority(info) {
        const { tokens, factor } = this;
        let priority = 2;
        for (let i = 0; i < tokens.length; i++) {
            const requestAmount = amountOrMinAmount(tokens[i]);
            if (info.deposits[i] < _factorAmount(factor, requestAmount)) {
                priority = 1;
                if (info.deposits[i] < requestAmount) {
                    priority = 0;
                    break;
                }
            }
        }
        return priority;
    }
    sort() {
        return this.arr.sort((a, b) => this._sortPriority(b) - this._sortPriority(a));
    }
}
// Simple factor multiplication; assumes request amounts stay within safe Number range in tests
export function _factorAmount(factor, amount) {
    return BigInt(Math.ceil(Number(amount) * factor));
}
export async function getSolventXlps(chainId, paymasters, mcAssets, includeBalance = true, offset = 0, length = 1000) {
    const paymasterAddress = paymasters.addressOn(chainId);
    const client = paymasters.client.clientOn(chainId);
    const assets = mcAssets.map((asset) => ({
        erc20Token: toAddress(chainId, asset.token),
        amount: amountOrMinAmount(asset)
    }));
    const helper = getXlpSelectionContract(client);
    const code = await client.getCode({ address: helper.address });
    if (code == null || code === '0x') {
        throw new Error(`XlpSelectionHelper not deployed on chain ${chainId} at address ${helper.address}`);
    }
    const result = await helper.read.getSolventXlps([paymasterAddress, offset, length, assets, includeBalance]);
    return result;
}
export function getXlpSelectionContract(client) {
    return getContract({
        client,
        address: getXlpSelectionHelperAddress(),
        abi: XlpSelectionHelperMeta.abi
    });
}
export function getXlpSelectionHelperAddress() {
    return getCreate2DeployAddress(XlpSelectionHelperMeta);
}
