// @eil-protocol/sdk/dist/utils/pollEvents.js
import { decodeEventLog, toEventSelector } from 'viem';
export class EventsPoller {
    eventAbis;
    topics;
    stopped = false;
    constructor(params) {
        const { abi, eventNames } = params;
        this.eventAbis = abi.filter((x) => x.type === 'event' && eventNames.includes(x.name));
        if (this.eventAbis.length != eventNames.length) {
            throw new Error(`Event names "${eventNames}" not found in contract ABI.`);
        }
        this.topics = this.eventAbis.map((abi) => toEventSelector(abi));
        void this.pollEvents(params).catch((e) => {
            console.error('pollEvents:', e);
        });
    }
    /**
     * run the poll event loop.
     * this method will poll for events on new blocks, and call the onLog function.
     * it only exists when stopEventPoller() is called
     */
    async pollEvents(params) {
        const { client, eventNames, onLog, pollInterval = 1000, fromBlock, } = params;
        let lastBlock = fromBlock ?? (await client.getBlockNumber()) - 1n;
        let interval = 200;
        while (!this.stopped) {
            await new Promise((r) => setTimeout(r, interval));
            interval = Math.min(interval * 2, pollInterval);
            const latestBlock = await client.getBlockNumber();
            if (latestBlock <= lastBlock)
                continue;
            let fromBlock = lastBlock + 1n;
            let toBlock = latestBlock;
            const logs = await client.getLogs({
                fromBlock,
                toBlock,
            });
            for (const log of logs) {
                try {
                    if (log.topics[0] == null || !this.topics.includes(log.topics[0])) {
                        continue;
                    }
                    const decoded = decodeEventLog({
                        abi: this.eventAbis,
                        data: log.data,
                        topics: log.topics,
                    });
                    if (!eventNames.includes(decoded.eventName)) {
                        continue;
                    }
                    await onLog({
                        ...log,
                        args: decoded.args,
                        eventName: decoded.eventName,
                    });
                }
                catch {
                    console.warn(`${eventNames}: ignored event topic`, log.topics[0]);
                    // ignore logs that don't match this event
                }
            }
            lastBlock = latestBlock;
        }
    }
    /**
     * stop the pollEvents
     */
    stopEventPoller() {
        this.stopped = true;
    }
}
