// @eil-protocol/sdk/dist/sdk/types/FunctionCall.js
export function getFunctionCallAbi(call) {
    const abi = call.abi ?? call.target?.abi;
    if (!abi) {
        throw new Error(`no ABI found for ${call}`);
    }
    return abi;
}
export function getFunctionCallAddress(call, chainId) {
    if (typeof call.target == 'string') {
        return call.target;
    }
    if ('addressOn' in call.target) {
        return call.target.addressOn(chainId);
    }
    else if ('address' in call.target) {
        return call.target.address;
    }
    else {
        throw new Error(`FunctionCall target is not a valid address: ${call.target}`);
    }
}
/**
 * @notice Determines if a parameter is a low-level encoded {@link Call} object or a {@link FunctionCall}.
 * @param call The call object to check.
 * @return True if the parameter is a Call object, false if it's a FunctionCall.
 * @typeGuard This is a type guard function that asserts the parameter is of the type {@link Call}.
 */
export function isCall(call) {
    return 'to' in call && 'data' in call;
}
/**
 * @notice Determines if an object is of the type {@link FunctionCall}.
 * @param call The call object to check.
 * @return True if the parameter is a FunctionCall object.
 * @typeGuard This is a type guard function that asserts the parameter is of the type {@link FunctionCall}.
 */
export function isFunctionCall(call) {
    return call.target != null && call.functionName != null;
}
/**
 * @notice Determines if a call has dynamic arguments.
 * @param call The call object to check.
 * @return True if at least one argument is dynamic (of type FunctionCall).
 * @typeGuard This is a type guard function that asserts the parameter is of type {@link FunctionCall}.
 */
export function hasDynamicArguments(call) {
    if (!isFunctionCall(call)) {
        return false;
    }
    return call.args?.some((arg) => isFunctionCall(arg)) ?? false;
}
