// @eil-protocol/sdk/dist/sdk/types/MultichainClient.js
//wrapper for PublicClient:
// the network "provider" for all blockchain read operations.
import { decodeFunctionResult, encodeFunctionData, toHex } from "viem";
export class MultichainClient {
    clients = new Map();
    static create(chains) {
        return new MultichainClient(chains.map(chain => chain.publicClient));
    }
    constructor(publicClients = []) {
        for (const client of publicClients) {
            this.addClientWithChainId(client, BigInt(client.chain.id));
        }
    }
    //return all clients, to do a "forEach"
    all() {
        return Array.from(this.clients.values());
    }
    async addClient(client) {
        const chainId = BigInt(await client.getChainId());
        this.addClientWithChainId(client, chainId);
    }
    addClientWithChainId(client, chainId) {
        if (this.clients.has(chainId)) {
            throw new Error(`Client already exists for chainId: ${chainId}`);
        }
        this.clients.set(chainId, client);
    }
    clientOn(chainId) {
        if (!this.clients.has(chainId)) {
            throw new Error(`No client found for chainId: ${chainId}. Supported chains: ${Array.from(this.clients.keys()).join(', ')}`);
        }
        return this.clients.get(chainId);
    }
    async call({ chainId, to, abi, functionName, args, value }) {
        const client = this.clientOn(chainId);
        const data = encodeFunctionData({ abi, functionName, args });
        const valueHex = value ? toHex(value) : undefined;
        const ret = await client.request({ method: 'eth_call', params: [{ to, data, value: valueHex }, 'latest'] });
        return decodeFunctionResult({ abi, functionName, data: ret });
    }
}
