// @eil-protocol/sdk/dist/sdk/types/MultichainContract.js
//todo: make this type-safe, using TAbi...
import { decodeFunctionResult, encodeFunctionData, isHex } from 'viem';
export class MultichainContract {
    client;
    abi;
    deployments = new Map();
    defaultAddr;
    static isMultichainContract(obj) {
        return typeof obj.addressOn == 'function';
    }
    constructor(client, abi, deployments = []) {
        this.client = client;
        this.abi = abi;
        if (!Array.isArray(deployments)) {
            if (!isHex(deployments)) {
                throw new Error(`deployment address is either array or specific address`);
            }
            this.defaultAddr = deployments;
        }
        else {
            for (const entry of deployments) {
                if (Array.isArray(entry)) {
                    const [chainId, address] = entry;
                    this.addAddress(BigInt(chainId), address);
                }
                else {
                    const { chainId, address } = entry;
                    this.addAddress(BigInt(chainId), address);
                }
            }
        }
    }
    addAddress(chainId, address) {
        if (this.deployments.has(chainId)) {
            throw new Error(`Address already exists for chainId: ${chainId}`);
        }
        this.deployments.set(chainId, address);
    }
    hasAddress(chainId) {
        return this.deployments.has(chainId);
    }
    addressOn(chainId) {
        if (!this.deployments.has(chainId) && !this.defaultAddr) {
            throw new Error(`MultiChainEntity does not have an address on chain ${chainId}`);
        }
        return (this.deployments.get(chainId) ?? this.defaultAddr);
    }
    encodeFunctionData(functionName, args) {
        return encodeFunctionData({ abi: this.abi, functionName, args });
    }
    decodeFunctionResult(functionName, data) {
        return decodeFunctionResult({ abi: this.abi, functionName, data });
    }
    async call(chainId, functionName, args, callOptions = {}) {
        const client = this.client.clientOn(chainId);
        const data = this.encodeFunctionData(functionName, args);
        const ret = await client.call({
            to: this.addressOn(chainId),
            data,
            ...callOptions
        });
        return this.decodeFunctionResult(functionName, ret.data);
    }
}
