// @eil-protocol/sdk/dist/sdk/types/MultichainToken.js
import { MultichainContract } from './MultichainContract.js';
import { erc20Abi } from 'viem';
import { toAddress } from './index.js';
//wrapper for multichain ERC20 token operations.
export class MultichainToken extends MultichainContract {
    constructor(name, client, deployments) {
        super(client, erc20Abi, deployments);
    }
    async balanceOf(chainId, address) {
        return this.call(chainId, 'balanceOf', [toAddress(chainId, address)]);
    }
    async symbol(chainId) {
        return this.call(chainId, 'symbol', []);
    }
    async decimals(chainId) {
        return this.call(chainId, 'decimals', []);
    }
    async allowance(chainId, owner, spender) {
        return this.call(chainId, 'allowance', [toAddress(chainId, owner), toAddress(chainId, spender)]);
    }
    async totalBalanceOf(targetAddress) {
        const balances = await Promise.all(this.client.all().map((client) => {
            const chainId = BigInt(client.chain.id);
            return this.balanceOf(chainId, targetAddress);
        }));
        const totalBalance = balances.reduce((acc, balance) => acc + balance);
        const perChainBalance = this.client.all().map((client, index) => {
            return { chainId: BigInt(client.chain.id), balance: balances[index] };
        });
        return {
            perChainBalance,
            totalBalance
        };
    }
}
