// @eil-protocol/sdk/dist/sdk/account/accountUtils.js
import { encodeAbiParameters, parseAbiParameters } from 'viem';
import { getUserOpHash } from '../sdkUtils/index.js';
const DOMAIN_NAME = 'CrossChainAccount';
const DOMAIN_VERSION = '1';
export async function oneSignatureSignUserOps(owner, userOps) {
    const userOpHashes = userOps.map(op => getUserOpHash(op));
    const sig = await owner.signTypedData({
        domain: getCrossChainTypedDataDomain(),
        types: getCrossChainTypedDataTypes(),
        primaryType: 'CrossChainAccount',
        message: { userOpHashes }
    });
    let signedUserOps = userOps.map((op, index) => {
        return {
            ...op,
            signature: encodeAbiParameters(parseAbiParameters('uint256, bytes, bytes32[]'), [BigInt(index), sig, userOpHashes])
        };
    });
    return signedUserOps;
}
export function getCrossChainTypedDataDomain() {
    return {
        name: DOMAIN_NAME,
        version: DOMAIN_VERSION
    };
}
export function getCrossChainTypedDataTypes() {
    return {
        CrossChainAccount: [
            { name: 'userOpHashes', type: 'bytes32[]' }
        ]
    };
}
