// @eil-protocol/sdk/dist/sdk/account/BaseMultichainSmartAccount.js
import { isFunctionCall } from '../types/index.js';
import { asCall } from '../sdkUtils/asCall.js';
/**
 * @title BaseMultichainSmartAccount
 * @notice Base class for managing cross-chain smart account instances.
 * @dev This abstract class implements cross-chain account functionality,
 * by hosting multiple {@link SmartAccount} instances on multiple chains.
 * Subclasses must implement:
 * - signUserOps(): Signs UserOps array for multichain execution.
 *                  For example, a basic implementation iterates accounts per-chain and requests individual signatures.
 *                  This is not ideal for UX as the user needs to confirm each signature separately.
 * - encodeCalls(): Encodes function calls array into a single UserOp execution.
 *                  This function supports dynamic {@link FunctionCall} parameters.
 */
export class BaseMultichainSmartAccount {
    bundlerManager;
    constructor(bundlerManager) {
        this.bundlerManager = bundlerManager;
    }
    addressOn(chainId) {
        return this.contractOn(chainId).address;
    }
    /**
     * Encode a batch of calls that are known to be non-dynamic for execution.
     * @dev The given array of {@link FunctionCall} objects must be all deterministic with no dynamic runtime variables.
     * @param chainId The chain ID of a chain this array of calls is intended for.
     * @param calls An array of static function calls to encode.
     * @return An encoded batch of calls as a hex string.
     * @throws {Error} If any of the calls are dynamic.
     */
    encodeStaticCalls(chainId, calls) {
        const hasDynamic = calls.some((call) => isFunctionCall(call));
        if (hasDynamic) {
            throw new Error('encodeStaticCalls: all calls must be static');
        }
        const account = this.contractOn(chainId);
        return account.encodeCalls(calls.map(call => asCall(chainId, call)));
    }
    async sendUserOperation(userOp) {
        return this.bundlerManager.sendUserOperation(userOp);
    }
    async verifyBundlerConfig(chainId, entryPoints) {
        await this.bundlerManager.verifyConfig(chainId, entryPoints);
    }
}
