// @eil-protocol/sdk/dist/sdk/builder/BatchBuilder.js
import { publicActions } from 'viem';
import { FunctionCallAction, prepareCallWithRuntimeVars, VoucherRequestAction } from '../actions/index.js';
import { isCall, isValidAddress, toAddress } from '../types/index.js';
import { appendPaymasterSignature, getUserOpHash } from '../index.js';
import { assert } from '../sdkUtils/SdkUtils.js';
import { abiEncodePaymasterData, nowSeconds } from '../../utils/index.js';
/**
 * Return the minimum amount for an asset.
 * If the amount value is fixed, return it as-is.
 * If the value is dynamic (determined on-chain), return the {@link minProviderDeposit} instead.
 * If {@link minProviderDeposit} is not defined, returns `1`.
 */
export function amountOrMinAmount(asset) {
    if (typeof asset.amount === 'bigint') {
        return asset.amount;
    }
    if (asset.minProviderDeposit == undefined) {
        return 1n; // real minimum unknown. just verify a non-zero value.
    }
    return asset.minProviderDeposit;
}
/**
 * BatchBuilder class - manages single chain operations within a {@link CrossChainBuilder}.
 * Contains all the actions to be executed on a specific chain by the Smart Account.
 * Handles the cross-chain token transfers via {@link SdkVoucherRequest}.
 */
export class BatchBuilder {
    parentBuilder;
    ephemeralSigner;
    coordinator;
    config;
    paymaster;
    chainId;
    isBuilt = false;
    /** List of input voucher requests to collect before executing all actions */
    inputVoucherRequests = [];
    /** List of source actions in this UserOperation */
    actions = [];
    /** Override fields when building this UserOperation (e.g., specific paymaster or gas values) */
    userOpOverrides;
    /** Names of all dynamic runtime variables used in this batch */
    _vars = new Set();
    constructor(parentBuilder, ephemeralSigner, coordinator, config, paymaster, chainId) {
        this.parentBuilder = parentBuilder;
        this.ephemeralSigner = ephemeralSigner;
        this.coordinator = coordinator;
        this.config = config;
        this.paymaster = paymaster;
        this.chainId = chainId;
    }
    endBatch() {
        return this.parentBuilder;
    }
    /**
     * Add a new dynamic runtime variable to the batch.
     * Variables can be used to store values that are not known at the time of batch creation.
     * Their actual values are determined on-chain at runtime.
     * For example, one action may need to operate on an outcome of an on-chain swap done as part of a single batch.
     * @param varName The name of the variable to be added to the batch.
     */
    addVar(varName) {
        this._vars.add(varName);
    }
    /**
     * Check if the batch contains a specific runtime variable.
     */
    hasVar(varName) {
        return this._vars.has(varName);
    }
    /**
     * Adds an action to the current batch. Actions are grouped by chainId.
     * Special actions (voucherRequest, useVoucherRequest) are used to manage flow between chains.
     * @param action The action to be added to the current batch
     * @returns this BatchBuilder instance for chaining
     */
    addAction(action) {
        this.assertNotBuilt();
        if (action instanceof VoucherRequestAction) {
            this.processVoucherRequest(action.voucherRequest);
        }
        if (action instanceof FunctionCallAction) {
            if (!isValidAddress(this.chainId, action.call.target)) {
                throw new Error(`Calling "${action.call.functionName}" on contract with no address on chain ${this.chainId}`);
            }
        }
        this.actions.push(action);
        return this;
    }
    /**
     * Overrides values in the generated UserOperation.
     * Use with caution as incorrect modifications may break the batch functionality.
     * @param overrides a partial {@link UserOperation} object containing fields to override.
     * @returns this {@link BatchBuilder} instance for chaining.
     */
    overrideUserOp(overrides) {
        this.assertNotBuilt();
        // should not override paymaster if the batch has inputs (depends on other batches)
        if (this.inputVoucherRequests.length > 0 && (overrides.paymaster || overrides.paymasterData)) {
            throw new Error(`Cannot override paymaster or paymasterData in a batch`);
        }
        this.userOpOverrides = overrides;
        return this;
    }
    processVoucherRequest(req) {
        if (req.destinationChainId == this.chainId) {
            throw new Error(`destinationChainId must be different than current chainId ${this.chainId}`);
        }
        if (req.sourceChainId == undefined) {
            req.sourceChainId = this.chainId;
        }
        else {
            assert(req.sourceChainId == this.chainId, `Voucher request sourceChainId ${req.sourceChainId} does not match batch chainId ${this.chainId}`);
        }
        assert(!this.coordinator.has(req), `Voucher request ${req.ref} already exists in this BatchBuilder`);
        req.tokens.forEach(token => {
            if (!isValidAddress(req.sourceChainId, token.token)) {
                throw new Error(`Voucher token does not have address on chain ${req.sourceChainId}`);
            }
            if (!isValidAddress(req.destinationChainId, token.token)) {
                throw new Error(`Voucher token does not have address on chain ${req.destinationChainId}`);
            }
        });
        this.coordinator.set({
            voucher: req,
            sourceBatch: this,
        });
    }
    /**
     * Add voucher creation to this batch.
     * This will trigger a transfer of the specified tokens to the destination chain.
     */
    addVoucherRequest(req) {
        return this.addAction(new VoucherRequestAction(req));
    }
    /**
     * Use the {@link SdkVoucherRequest} created in an earlier batch to move tokens to this chain.
     */
    useVoucher(refId) {
        this.assertNotBuilt();
        const internalVoucherInfo = this.coordinator.getVoucherInternalInfo(refId);
        const voucher = internalVoucherInfo?.voucher;
        assert(internalVoucherInfo != null, `Voucher request ${refId} not found in action builder`);
        assert(this.userOpOverrides?.paymaster == null && this.userOpOverrides?.paymasterData == null, `Cannot override paymaster or paymasterData in a batch that uses vouchers.`);
        assert(internalVoucherInfo.destBatch == undefined, `Voucher request ${refId} already used`);
        assert(voucher.destinationChainId == this.chainId, `Voucher request is for chain ${voucher.destinationChainId}, but batch is for chain ${this.chainId}`);
        internalVoucherInfo.destBatch = this;
        this.inputVoucherRequests.push(voucher);
        return this;
    }
    /**
     * Use all voucher requests for this chain from all previous batches.
     * Throws if no voucher requests were added with the current chain as the destination.
     */
    useAllVouchers() {
        let added = false;
        const allVoucherRequests = this.coordinator.getAllOutVoucherRequests();
        for (const v of allVoucherRequests) {
            if (v.destinationChainId === this.chainId) {
                added = true;
                this.useVoucher(v.ref);
            }
        }
        assert(added, `No voucher requests found for chain ${this.chainId}`);
        return this;
    }
    /**
     * Create a {@link UserOperation} with the provided {@link SingleChainBatch} calls and {@link SdkVoucherRequest} list.
     * Notice that all Voucher Requests are appended to the end of the batch regardless of the order they were provided.
     * @returns - a {@link Partial} of the {@link UserOperation}.
     */
    async createUserOp() {
        const chainId = this.chainId;
        const mcAccount = this.parentBuilder.getAccount();
        const smartAccount = mcAccount.contractOn(chainId);
        const allCalls = await Promise.all(this.actions.map((action) => {
            return action.encodeCall(this);
        }));
        const calls = allCalls.flat().map(call => {
            if (isCall(call)) {
                return call;
            }
            else {
                return prepareCallWithRuntimeVars(call, this);
            }
        });
        const [sender, nonce, { factory, factoryData }, callData] = await Promise.all([
            smartAccount.getAddress(),
            smartAccount.getNonce(),
            smartAccount.getFactoryArgs(),
            calls.length == 0 ? '0x' : mcAccount.encodeCalls(chainId, calls)
        ]);
        const nonce1 = BigInt(nowSeconds()) << 64n;
        const { maxFeePerGas, maxPriorityFeePerGas } = await smartAccount.client.extend(publicActions).estimateFeesPerGas();
        let userOp = {
            chainId,
            sender,
            nonce: nonce1,
            factory,
            factoryData,
            callData,
            callGasLimit: 3000000n,
            verificationGasLimit: 500000n,
            preVerificationGas: 100000n,
            maxFeePerGas,
            maxPriorityFeePerGas,
            ...this.userOpOverrides
        };
        const { paymaster, paymasterData, paymasterVerificationGasLimit, paymasterPostOpGasLimit } = await this.createPaymasterInfo(userOp, this.inputVoucherRequests);
        if (paymaster != null) {
            userOp = {
                ...userOp,
                paymaster,
                paymasterData,
                paymasterVerificationGasLimit,
                paymasterPostOpGasLimit
            };
        }
        return userOp;
    }
    async createPaymasterInfo(userOp, voucherRequests) {
        if (voucherRequests.length === 0) {
            if (this.config.input.sourcePaymaster != null) {
                // use source paymaster
                return this.config.input.sourcePaymaster.getPaymasterStubData(userOp);
            }
            else {
                // no paymaster. account pays with eth.
                return {};
            }
        }
        return {
            paymaster: this.paymaster,
            paymasterData: this._createPaymasterData(voucherRequests),
            paymasterVerificationGasLimit: 500000n, //todo: estimate gas
            paymasterPostOpGasLimit: 100000n
        };
    }
    _createPaymasterData(voucherRequests) {
        if (voucherRequests.length === 0) {
            // no voucher requests meaning no paymaster data for the AtomicSwapPaymaster
            return '0x';
        }
        // encode the voucher requests into the paymaster data
        const basePaymasterData = abiEncodePaymasterData({
            vouchersAssetsMinimums: voucherRequests.map(this._toVouchersAssetsMinimums.bind(this)),
            ephemeralSigner: this.ephemeralSigner.address
        });
        return appendPaymasterSignature(basePaymasterData, '0x01');
    }
    _toVouchersAssetsMinimums(action) {
        const chainId = action.destinationChainId;
        return action.tokens.map(asset => ({
            erc20Token: toAddress(chainId, asset.token),
            amount: amountOrMinAmount(asset)
        }));
    }
    async buildSingleChainBatch() {
        const userOp = await this.createUserOp();
        const userOpHash = getUserOpHash(userOp);
        return {
            userOp,
            userOpHash,
            chainId: this.chainId,
            inputVoucherRequests: this.inputVoucherRequests,
            outVoucherRequests: this.getOutVoucherRequests(),
        };
    }
    getVoucherInternalInfo(voucher) {
        return this.coordinator.getVoucherInternalInfo(voucher.ref);
    }
    getOutVoucherRequests() {
        return this.actions
            .filter(action => action instanceof VoucherRequestAction)
            .map(action => action.voucherRequest);
    }
    assertNotBuilt() {
        if (this.isBuilt) {
            throw new Error('CrossChainBuilder already built. Create a new instance to build a new session.');
        }
    }
}
