// @eil-protocol/sdk/dist/sdk/builder/CrossChainVoucherCoordinator.js
export class CrossChainVoucherCoordinator {
    /**
     * A mapping from the inputs of the Builder to the full voucher requests information for all batches.
     */
    vouchersInternalInfo = new Map();
    getAllVoucherInternalInfos() {
        return [...this.vouchersInternalInfo.values()];
    }
    getVoucherInternalInfo(refId) {
        return this.vouchersInternalInfo.get(refId);
    }
    getAllOutVoucherRequests() {
        return Array.from(this.vouchersInternalInfo.values()).map(info => info.voucher);
    }
    has(sdkVoucherRequest) {
        return this.vouchersInternalInfo.has(sdkVoucherRequest.ref);
    }
    set(internalVoucherInfo) {
        this.vouchersInternalInfo.set(internalVoucherInfo.voucher.ref, internalVoucherInfo);
    }
    updateVoucherXlps(voucherReq, xlps) {
        const info = this.vouchersInternalInfo.get(voucherReq.ref);
        if (!info) {
            throw new Error(`Voucher request ${voucherReq.ref} not found in action builder`);
        }
        info.allowedXlps = xlps;
    }
}
