// @eil-protocol/sdk/dist/sdk/builder/MultichainBundlerManager.js
import { createClient, fromHex, http } from 'viem';
import { stringifyBigIntReplacer } from '../sdkUtils/SdkUtils.js';
export class MultichainBundlerManager {
    chainInfos;
    constructor(chainInfos) {
        this.chainInfos = chainInfos;
        for (const { chainId, bundlerUrl, entryPointAddress } of chainInfos) {
            let provider = createClient({
                transport: http(bundlerUrl, { retryCount: 0 })
            }).transport;
            if (bundlerUrl !== undefined) {
                this.addBundler(chainId, provider, entryPointAddress);
            }
        }
    }
    bundlers = new Map();
    entryPoints = new Map();
    addBundler(chainId, provider, entryPointAddress) {
        this.bundlers.set(chainId, provider);
        this.entryPoints.set(chainId, entryPointAddress);
    }
    async verifyConfig(chainId, entryPoint) {
        const bundler = this.bundlers.get(chainId);
        if (bundler == null) {
            throw new Error(`No bundler configured for chain ${chainId}`);
        }
        const bundlerChainId = fromHex(await bundler.request({ 'method': 'eth_chainId' }), 'bigint');
        if (bundlerChainId != chainId) {
            throw new Error(`Bundler for chain ${chainId} is configured with a wrong chainId ${bundlerChainId}`);
        }
        const supportedEntryPoints = await bundler.request({ 'method': 'eth_supportedEntryPoints' });
        if (!supportedEntryPoints.map(addr => addr.toLowerCase()).includes(entryPoint.toLowerCase())) {
            throw new Error(`Bundler for chain ${chainId} does not support EntryPoint at address ${entryPoint}, only ${supportedEntryPoints}`);
        }
    }
    sendUserOperation(userOp) {
        const provider = this.bundlers.get(userOp.chainId);
        if (!provider) {
            throw new Error(`No bundler found for chainId: ${userOp.chainId}`);
        }
        const jsonUserOp = JSON.parse(JSON.stringify(userOp, stringifyBigIntReplacer));
        return provider.request({
            method: 'eth_sendUserOperation',
            params: [jsonUserOp, userOp.entryPointAddress]
        });
    }
}
