// @eil-protocol/sdk/dist/contractTypes/Voucher.js
import { encodeFunctionData, maxUint256, sliceHex, zeroAddress } from 'viem';
import { AtomicSwapTypes } from '../abitypes/abiTypes.js';
import { getVoucherRequestId, addressOf } from '../utils/index.js';
import { VoucherType } from './VoucherType.js';
export const EmptyVoucher = {
    voucherRequestDest: {
        chainId: 0n,
        sender: zeroAddress,
        paymaster: zeroAddress,
        assets: [],
        maxUserOpCost: 0n,
        expiresAt: 0n
    },
    requestId: `0x${'00'.repeat(32)}`,
    originationXlpAddress: zeroAddress,
    expiresAt: maxUint256,
    xlpSignature: '0x',
    voucherType: VoucherType.STANDARD
};
export async function signAtomicSwapVoucher(signer, voucher, voucherRequest) {
    const message = encodeForSigning(voucher, voucherRequest);
    return await signer.signMessage({ message: { raw: message } });
}
export function encodeForSigning(voucher, voucherRequest) {
    const data = encodeFunctionData({
        abi: AtomicSwapTypes.abi,
        functionName: 'getDataForVoucherSignature',
        args: [voucherRequest.destination, voucher.requestId, voucher.originationXlpAddress, voucher.expiresAt, voucher.voucherType]
    });
    return sliceHex(data, 4); // remove selector
}
export async function getVoucherFromRequest(voucherRequest, xlpEOA, options = {}) {
    const requestId = getVoucherRequestId(voucherRequest);
    const xlpAddressEOA = addressOf(xlpEOA);
    const voucher = {
        requestId,
        voucherRequestDest: voucherRequest.destination,
        originationXlpAddress: options.originationXlpAddress ?? xlpAddressEOA,
        expiresAt: options.expiresAt ?? voucherRequest.destination.expiresAt,
        xlpSignature: '0x',
        voucherType: options.voucherType ?? (options.isVoucherOverride ? VoucherType.OVERRIDE : VoucherType.STANDARD)
    };
    voucher.xlpSignature = await signAtomicSwapVoucher(xlpEOA, voucher, voucherRequest);
    return voucher;
}
