// @eil-protocol/sdk/dist/sdk/sdkUtils/UserOperationUtils.js
import { concatHex, hexToBigInt, size, sliceHex, toHex } from 'viem';
import { getUserOperationHash as viemGetUserOperationHash } from 'viem/account-abstraction';
const PAYMASTER_SIG_MAGIC = '0x22e325a297439656';
const PAYMASTER_SIG_MAGIC_LEN = 8;
const PAYMASTER_SIGNATURE_TRAILER_LEN = PAYMASTER_SIG_MAGIC_LEN + 2; // uint16 length + magic
export function getUserOpHash(op) {
    return viemGetUserOperationHash({
        userOperation: withSanitizedPaymasterData(op),
        entryPointAddress: op.entryPointAddress,
        entryPointVersion: '0.8',
        chainId: Number(op.chainId)
    });
}
export function appendPaymasterSignature(paymasterData, paymasterSignature) {
    const base = stripPaymasterSignature(paymasterData ?? '0x');
    const signature = (paymasterSignature ?? '0x');
    const sigLen = size(signature);
    const length2bytes = toHex(sigLen, { size: 2 });
    return concatHex([
        base,
        signature,
        length2bytes,
        PAYMASTER_SIG_MAGIC
    ]);
}
export function getPaymasterDataForSigning(paymasterAndData) {
    const pmSigLength = getPaymasterSignatureLength(paymasterAndData);
    if (pmSigLength === undefined) {
        return paymasterAndData;
    }
    const dataLength = size(paymasterAndData);
    return concatHex([
        sliceHex(paymasterAndData, 0, dataLength - pmSigLength - PAYMASTER_SIGNATURE_TRAILER_LEN),
        PAYMASTER_SIG_MAGIC
    ]);
}
export const maxUint48 = (2 ** 48) - 1;
export function parseValidationData(validationData) {
    const data = toHex(validationData, { size: 32 });
    const aggregator = data.slice(-40);
    let validUntil = parseInt(data.slice(-52, -40), 16);
    if (validUntil === 0) {
        validUntil = maxUint48;
    }
    const validAfter = parseInt(data.slice(2, 14), 16);
    return {
        aggregator,
        validAfter,
        validUntil
    };
}
function withSanitizedPaymasterData(op) {
    return {
        ...op,
        paymasterData: getPaymasterDataForSigning(op.paymasterData ?? '0x')
    };
}
function getPaymasterSignatureLength(paymasterAndData) {
    const dataLength = size(paymasterAndData);
    const suffixLength = PAYMASTER_SIG_MAGIC_LEN;
    if (dataLength > suffixLength &&
        sliceHex(paymasterAndData, dataLength - suffixLength).toLowerCase() === PAYMASTER_SIG_MAGIC) {
        return Number(hexToBigInt(sliceHex(paymasterAndData, dataLength - PAYMASTER_SIGNATURE_TRAILER_LEN, dataLength - suffixLength)));
    }
    else {
        return undefined;
    }
}
function stripPaymasterSignature(paymasterData) {
    const pmSigLen = getPaymasterSignatureLength(paymasterData);
    if (pmSigLen === undefined) {
        return paymasterData;
    }
    const dataLength = size(paymasterData);
    return sliceHex(paymasterData, 0, dataLength - pmSigLen - PAYMASTER_SIGNATURE_TRAILER_LEN);
}
